src/workerd/api/crypto-impl-asymmetric.c++

```diff
@@ -1814,9 +1813,63 @@ public:
     return !!result;
   }
 
+  kj::Array<kj::byte> deriveBits(
+      SubtleCrypto::DeriveKeyAlgorithm&& algorithm,
+      kj::Maybe<uint32_t> resultBitLength) const override final {
+    JSG_REQUIRE(getAlgorithmName() == "X25519", DOMNotSupportedError, ""
+        "The deriveBits operation is not implemented for \"", getAlgorithmName(), "\".");
+
+    JSG_REQUIRE(getType() == "private"_kj, DOMInvalidAccessError, ""
+        "The deriveBits operation is only valid for a private key, not \"", getType(), "\".");
+
+    auto& publicKey = JSG_REQUIRE_NONNULL(algorithm.$public, TypeError,
+        "Missing field \"public\" in \"derivedKeyParams\".");
+
+    JSG_REQUIRE(publicKey->getType() == "public"_kj, DOMInvalidAccessError, ""
+        "The public key provided has type \"", publicKey->getType(), "\", not \"public\"");
+
+    JSG_REQUIRE(getAlgorithm().which() == publicKey->getAlgorithm().which(), DOMInvalidAccessError,
+        "Base ", getAlgorithmName(), " private key cannot be used to derive a "
+        "key from a peer ", getAlgorithmName(), " public key");
+
+    JSG_REQUIRE(getAlgorithmName() == publicKey->getAlgorithmName(), DOMInvalidAccessError,
+        "Private key for derivation is using \"", getAlgorithmName(),
+        "\" while public key is using \"", publicKey->getAlgorithmName(), "\".");
+
+    // The check above for the algorithm `which` equality ensures that the impl can be downcast to
+    // EdDsaKey (assuming we don't accidentally create a class that doesn't inherit this one that
+    // for some reason returns an EdDsaKey).
+    auto& publicKeyImpl = kj::downcast<EdDsaKeyBase>(*publicKey->impl);
+
+    // EDDH code derived from https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_derive.html
+    auto ctx = OSSL_NEW(EVP_PKEY_CTX, getEvpPkey(), nullptr);
+    JSG_REQUIRE(1 == EVP_PKEY_derive_init(ctx), InternalDOMOperationError,
+        "Failed to init EDDH key derivation", internalDescribeOpensslErrors());
+    JSG_REQUIRE(1 == EVP_PKEY_derive_set_peer(ctx, publicKeyImpl.getEvpPkey()),
+        InternalDOMOperationError, "Failed to set EDDH peer", internalDescribeOpensslErrors());
+
+    kj::Vector<kj::byte> sharedSecret(X25519_SHARED_KEY_LEN);
+    size_t skeylen = 0;
+    JSG_REQUIRE(1 == EVP_PKEY_derive(ctx, NULL, &skeylen), DOMOperationError,
+        "Failed to derive EDDH key", internalDescribeOpensslErrors());
+    KJ_ASSERT(skeylen == X25519_SHARED_KEY_LEN);
+    JSG_REQUIRE(1 == EVP_PKEY_derive(ctx, sharedSecret.begin(), &skeylen), DOMOperationError,
+        "Failed to derive EDDH key", internalDescribeOpensslErrors());
+
+    // Check for all-zero value as mandated by spec
+    kj::byte isNonZeroSecret = 0;
+    for (kj::byte b : sharedSecret) {
+      isNonZeroSecret |= b;
+    }
+    JSG_REQUIRE(isNonZeroSecret == 0, DOMOperationError,
```

Comment:
This condition seems odd to me at first sight. It seems to me that `isNonZeroSecret == 0` if and only if all bytes of `sharedSecret` are 0, so shouldn't this assertion be negated? (I haven't looked at this codebase or the spec in a long time so I might be missing something.)

Reply:
You're right, the assertion is wrong here – I'll take care of this on Monday